concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+and so on.
+
+As with infinite recursives, note that it's still useful to do math around these.
+For example, if you have "a string or a null",
+you have a [sum type](#sum-types) (more specifically, a [maybe type](#maybe-types)),
+and even if the cardinality of string is "infinite", it's still useful to declare
+the cardinality of your string is "a" (an arbitrary placeholder),
+and then the cardinality of "string or null" is "|a+1|".
+(In practice: how often have you had an API bug burn you on the different
+between null and an empty string, or the difference between null and absent?
+This is why we keep counting cardinality even after an infinity shows up!)
+
+### unit types
+
+Let's talk about one last size of cardinality... "unit": unit just means "one".
+
+"Unit" types don't come up too often... or at least, they're not often named as such.
+Just like "product types" are quite common once you recognize them, though, so is "unit".
+
+"Null" can be thought of as a "unit" type.
+
+"Maybe"/"Optional" types (further discussion later in the [Types in the Wild](#types-in-the-wild) chapter,
+specifically the [maybe types](#maybe-types) section) can be thought of as "unit plus a thing".
+
+Enums can be considered a sum type of where every type is itself a just another named unit type!
+(Whether or not you find this a *useful* perspective is up to you.  If it's confusing, forget it.)
+
+
+Open vs Closed membership
+-------------------------
+
+All of the above [categories of types](#categories-of-types) have one thing in common:
+they have "closed" membership.  That is, at "compile" (or other static analysis) time,
+we know what their membership is, and we can completely count their cardinality (infinities notwithstanding).
+Once the program is compiled(/analyzed), there's no way to add new members to any of these types,
+and no reason to consider the idea of updating any of the cardinality accounting.
+
+What if we want "open" membership, where I *don't* know all of the possible
+types of values that will be part of a type in advance?
+
+This "open membership" is commonly known as **interfaces**.
+
+### interfaces
+
+Interfaces have the concept of "open membership", and are also typically
+combined with some definition of "features" that are required if something
+will be considered to implement the interface.
+
+In programming languages, the definition of "features" usually is expressed in
+terms of a set of methods.  This makes sense for programming languages,
+because if you don't know what methods you can call on a thing,
+you can't really do much with the thing!
+
+One could also imagine using a different definition of "features" to describe
+"interfaces" that apply purely to data:
+For example, stating that some map has the features to conform to an interface
+only if some entries of some known names and value types are present in the map.
+(However, this is not often seen, so this document will continue to frequently
```

Comment:
Maybe it's not often expressed in these terms, but this is exactly how "stringly typed interfaces" play out in practice. JSON APIs are an obvious example.

Reply:
Hm, maybe I should rewrite this paragraph somehow.  I'm just not sure quite how.

The concept of "interfaces but for data" is what I'm trying to get to in the very end, because that's the direction that IPLD Schemas explore into.

But I'm not actually sure if JSON APIs in practice are a great example for this.  Or rather, they are, but with a zillion fascinating caveats that make me not want to give it its own entire space rather than try to treat it in this document.  (Lots of JSON APIs in practice end up using some language's type facilities to define *sort of*/*most of* the rules, but end up suffering from this in at least three different ways: the language semantics intrude (e.g. nullability appearing even if you don't want a +1 cardinality); the language types don't contain *all* the rules; and those facts combining make a nightmare for documentation, clarity, compatibility, etc.)

And "stringly typed" invokes the idea of giving up on compile-time or useful automation around type systems entirely, and writing a bunch of ad-hoc stuff.  Whereas where I'd like to go is towards useful type systems and automation and away from ad-hoc stuff, even while we stay close to the realities of serial data.